src/main/java/frc/robot/subsystems/Drivetrain.java

```diff
@@ -1,65 +1,525 @@
+        Module.create(ModuleConstants.COMP_FL, m_swerveModulesTab),
+        Module.create(ModuleConstants.COMP_FR, m_swerveModulesTab),
+        Module.create(ModuleConstants.COMP_BL, m_swerveModulesTab),
+        Module.create(ModuleConstants.COMP_BR, m_swerveModulesTab)
+      };
+    m_prevModule = m_modules[0];
+
+    m_odometry = new SwerveDriveOdometry(m_kinematics, m_pigeon.getRotation2d(), getModulePositions(), m_robotPose);
+    m_rotationController.enableContinuousInput(-Math.PI,Math.PI);
+    DoubleSupplier[] poseSupplier = {() -> getPose().getX(), () -> getPose().getY(), () -> getPose().getRotation().getRadians()};
+    LogManager.addDoubleArray("Pose2d", poseSupplier);
+
+    m_fieldDisplay.setRobotPose(getPose());
+    SmartDashboard.putData("Field", m_fieldDisplay);
+    
+  }
+
+  @Override
+  public void periodic() {
+    if (!Robot.isReal()) {
+      for (int i = 0; i < m_modules.length; i++) {
+        m_modules[i].periodic();
+      }
+    }
+    updateOdometry();
+    
+    m_fieldDisplay.setRobotPose(getPose());
+  }
+
+  public void runChassisPID(double x, double y, double rot) {
+    double xSpeed = m_xController.calculate(m_odometry.getPoseMeters().getX(), x);
+    double ySpeed = m_yController.calculate(m_odometry.getPoseMeters().getY(), y);
+    double rotRadians = m_rotationController.calculate(getAngleHeading(), rot);
+    // System.out.println(rotRadians);
+    driveRot(xSpeed, ySpeed, rotRadians, true);
+  }
+
+  public void setPigeonYaw(double degrees) {
+    m_pigeon.setYaw(degrees);
   }
 
   /**
-   * Drives the robot using tank drive controls Tank drive is slightly easier to code but less
-   * intuitive to control, so this is here as an example for now
-   *
-   * @param leftPower the commanded power to the left motors
-   * @param rightPower the commanded power to the right motors
+   * Resets the pigeon yaw, but only if it hasn't already been reset. Will reset it to {@link DriveConstants.kStartingHeadingDegrees}
    */
-  public void tankDrive(double leftPower, double rightPower) {
-    m_leftMotor1.set(ControlMode.PercentOutput, leftPower);
-    m_rightMotor1.set(ControlMode.PercentOutput, rightPower);
+  public void initializePigeonYaw(boolean force) {
+    if (!m_hasResetYaw || force) {
+      m_hasResetYaw = true;
+      setPigeonYaw(DriveConstants.kStartingHeadingDegrees);
+    }
   }
 
   /**
-   * Drives the robot using arcade controls.
+   * Method to drive the robot using joystick info.
    *
-   * @param forward the commanded forward movement
-   * @param turn the commanded turn rotation
+   * @param xSpeed speed of the robot in the x direction (forward)
+   * @param ySpeed speed of the robot in the y direction (sideways)
+   * @param rot angular rate of the robot
+   * @param fieldRelative whether the provided x and y speeds are relative to the field
+   */
+  public void driveRot(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
+
+
+    m_swerveModuleStates =
```

Comment:
For simulation to work, there need to be pigeon yaw update somewhere around this code. It is probably OK leave it as is for now, if we are planning to have a separate PR from simulation branch.

Reply:
I intentialy removed most if not all of the simulation features as to have it on a sepreate PR so they will be added later